examples/github-deploy.yml

```
@@ -44,10 +44,10 @@ jobs:
   upload_pypi:
     needs: [build_wheels, build_sdist]
     runs-on: ubuntu-latest
-    # upload to PyPI on every tag starting with 'v'
-    if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags/v')
-    # alternatively, to publish when a GitHub Release is created, use the following rule:
-    # if: github.event_name == 'release' && github.event.action == 'published'
+    environment: pypi
+    permissions:
+      id-token: write
+    if: github.event_name == 'release' && github.event.action == 'published'
```

I'd rather keep the reference for release-on-tag, some projects don't bother with Github releases (e.g. django, pip, tornado). But they're in the minority, so switching the default to release-based makes sense

```suggestion
    environment: pypi
    permissions:
      id-token: write
    if: github.event_name == 'release' && github.event.action == 'published'
    # or, alternatively, upload to PyPI on every tag starting with 'v' (remove on: release above to use this)
    # if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags/v')
```

---

Do environments work properly with release on tag, @webknjaz? I was rather assuming things like review deployments might be tied to releases.

---

I see it mentioned in the docs for the action, so it must be okay. I think "environments" and "releases" are not really as integrated as I thought (which might be why I can make a release then still have to review the deployment).

---

@henryiii not sure why you thought they depend on the events.. I'm not aware of any problems. PyPI only checks the env name, and on the GH side you may have extra protections attached.

By the way, I like to use the full syntax and specify env URLs so they are better embedded in the UI.